### DISCUSSION TITLE
Mandatory traits

### DISCUSSION THREAD

**Discussion by zoran995** [TerriaJS/terriajs#5198] [Q&A]
When defining catalog items do we have a way to determine mandatory traits (that user has to define)? The main use case is the development of a generic configurator application and for example, WMS has over 70 possible options. But also documentation could benefit from something like this.

**Comment by tephenavies** [comment]
I had intentions to add this to Trait definitions for documentation. A good point raised to me by @na9da is that some traits are required in some models but not others. E.g. for most items the `url` trait is required, but not (among others) CSV & GeoJson which can define data in a string. We may need to have some configuration for this kind of thing external to the definition of Traits that can deal with that.

**Reply by zoran995** [reply]
good point from @na9da, this might get pretty much complex

**Reply by zoran995** [reply]
I was looking a little more into this and will share my initial findings, so others can drop their thoughts.
The first idea was to add another descriptor that could be used to make some trait optional or mandatory. I believe that we could use this to override trait behaviour after mixing traits, it would add more complexity to the trait system for sure and I am afraid that it might prove to be unmanageable and doesn't force us to implement it. Another huge benefit from having a list of mandatory traits would be to validate the provided catalog item and log the error (but that is a separate issue). I used something similar to provide the default value for URL in #5145. It could look something like this.
```
class UrlTraits {
    @primitiveTrait(...)
    @optional(false)
    url?: string;
}
```

```
class geojsonTraits {
    @optional(true)
    url?: string;
}
```
Another idea was to add this to CreateModel, but it seems that in the end, it would create more issues then benefits
